File: boss-call/bin/boss-call.mjs
#!/usr/bin/env node
/**
 * boss-call — one Boss, several members, one line each.
 *
 *   boss-call setup                      link the skill into every harness on this machine
 *   boss-call host <room>                become the boss of a room
 *   boss-call join <room> --as <name>    join a room as a member; the current directory is your repo
 *   boss-call who                        which side you are on
 *   boss-call read [--ack]               unread mail for you
 *   boss-call wait [--timeout 25]        block until mail arrives (acknowledged), then print it
 *                                        (--timeout 590 with a 600 s shell-tool timeout: fewer empty returns)
 *   boss-call post [--kind K] "text"     a member: to the boss; the boss: --to <member>|all
 *   boss-call status | tail [-n N]       the room at a glance
 *   boss-call serve [--once]             run unattended: mail -> one kiso turn -> status
 *   boss-call peek-session [id|latest]   where a kiso session stands, from its log
 *
 * Room and identity come from the current directory once you have joined;
 * --room / --me / BOSS_CALL_ROOM / BOSS_CALL_ME override.
 */
import { existsSync, lstatSync, mkdirSync, readlinkSync, rmSync, symlinkSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { HOME, BossCallError, ack, formatMessage, heartbeat, host, identify, joinRoom, loadRoom, post, readMessages, resolveRoom, status, statusNudge, unread, waitForMail } from "../src/mailbox.mjs";
import { peek } from "../src/peek.mjs";
import { serve } from "../src/serve.mjs";

const PKG = resolve(dirname(fileURLToPath(import.meta.url)), "..");

function parse(argv) {
	const flags = {};
	const pos = [];
	for (let i = 0; i < argv.length; i++) {
		const a = argv[i];
		if (a === "--") {
			pos.push(...argv.slice(i + 1));
			break;
		}
		if (a.startsWith("--")) {
			const [k, inline] = a.slice(2).split(/=(.*)/s);
			const next = argv[i + 1];
			if (inline !== undefined) flags[k] = inline;
			else if (next !== undefined && !next.startsWith("--")) {
				flags[k] = next;
				i++;
			} else flags[k] = true;
		} else if (a === "-n" && argv[i + 1]) {
			flags.n = argv[++i];
		} else pos.push(a);
	}
	return { flags, pos };
}

function die(msg) {
	console.error(`boss-call: ${msg}`);
	process.exit(2);
}

function numberFlag(value, fallback, label, { integer = false, min = 0 } = {}) {
	const n = Number(value ?? fallback);
	if (!Number.isFinite(n) || (integer && !Number.isInteger(n)) || n < min) die(`${label} must be ${integer ? "an integer" : "a number"} >= ${min}`);
	return n;
}

/**
 * Where harnesses look for skills. `~/.agents/skills` is the shared convention
 * (pi, opencode and others read it); Claude Code and Codex have their own
 * directories. boss-call uses only these skill links and the CLI.
 */
const SKILL_DIRS = [
	{ at: [".agents", "skills"], always: true, note: "pi, opencode, any agentskills.io harness" },
	{ at: [".claude", "skills"], note: "Claude Code" },
	{ at: [".codex", "skills"], note: "Codex" },
	{ at: [".kiso", "skills"], note: "kiso" },
];

function link(target, at) {
	mkdirSync(dirname(at), { recursive: true });
	let current;
	try {
		current = lstatSync(at);
	} catch (err) {
		if (err.code !== "ENOENT") throw err;
	}
	if (current?.isSymbolicLink() && readlinkSync(at) === target) return "kept";
	if (current?.isSymbolicLink()) rmSync(at);
	else if (current) throw new BossCallError(`refusing to replace existing non-symlink: ${at}`);
	symlinkSync(target, at);
	return "linked";
}

function onPath(name) {
	return (process.env.PATH ?? "").split(":").some((d) => d && existsSync(join(d, name)));
}

function cmdSetup() {
	const home = homedir();
	for (const d of SKILL_DIRS) {
		if (!d.always && !existsSync(join(home, d.at[0]))) continue;
		const r = link(PKG, join(home, ...d.at, "boss-call"));
		console.log(`  ${r.padEnd(6)} ~/${d.at.join("/")}/boss-call  (${d.note})`);
	}
	const oldExt = join(home, ".kiso", "extensions", "boss-call.mjs");
	try {
		if (lstatSync(oldExt).isSymbolicLink()) {
			rmSync(oldExt);
			console.log("  removed ~/.kiso/extensions/boss-call.mjs (legacy extension; boss-call is CLI + skills only)");
		}
	} catch {}
	if (!onPath("boss-call")) {
		const r = link(join(PKG, "bin", "boss-call.mjs"), join(home, ".local", "bin", "boss-call"));
		console.log(`  ${r.padEnd(6)} ~/.local/bin/boss-call  (add ~/.local/bin to PATH if it is not)`);
	}
	console.log("\nnext: the boss runs `boss-call host <room>`; each member runs `boss-call join <room> --as <name>` inside its repo.\nA session becomes a member only when the person says: follow the boss-call skill. Otherwise it is an ordinary session.");
}

function main(argv) {
	const { flags, pos } = parse(argv);
	const cmd = pos.shift();
	if (!cmd || cmd === "help" || flags.help) {
		console.log(import.meta.url ? readHelp() : "");
		return 0;
	}
	if (cmd === "setup") return cmdSetup(), 0;

	if (cmd === "host") {
		const room = pos[0] ?? flags.room ?? die("host <room>");
		const name = flags.as ?? "boss";
		const kiso = kisoOpts(flags);
		host(room, name, { root: flags.root ? resolve(flags.root) : undefined, kiso });
		console.log(`you are the boss of ${room} as ${name}${kiso ? ` (serve: kiso${kiso.profile ? ` --model ${kiso.profile}` : ""})` : ""}`);
		return cmdWho({ room, me: name });
	}
	if (cmd === "join") {
		const room = pos[0] ?? flags.room ?? die("join <room> --as <name>");
		const name = flags.as ?? die("join needs --as <name>");
		const root = resolve(flags.root ?? process.cwd());
		joinRoom(room, name, root, { kiso: kisoOpts(flags) });
		console.log(`joined ${room} as ${name}, repo ${root}`);
		return cmdWho({ room, me: name });
	}

	const room = resolveRoom(flags.room, flags.cwd);
	switch (cmd) {
		case "who":
			return cmdWho({ room, me: flags.me, cwd: flags.cwd });
		case "status": {
			const s = status(room);
			console.log(`room ${room}: boss=${s.boss ?? "-"}  ${s.total} messages`);
			for (const r of s.rows) console.log(`  ${r.role.padEnd(6)} ${r.name.padEnd(12)} unread=${String(r.unread).padStart(3)} asks=${String(r.asks).padStart(2)} ${presence(r.heartbeat).padEnd(22)} last-posted=${r.lastPosted ?? "-"}`);
			return 0;
		}
		case "tail": {
			const n = numberFlag(flags.n, 20, "-n", { integer: true });
			for (const m of readMessages(room).slice(-n)) console.log(`#${String(m.seq).padStart(4)} ${m.ts} ${m.from.padStart(9)} -> ${m.to.padEnd(9)} [${m.kind.padEnd(6)}] ${(m.text.split("\n")[0] ?? "").slice(0, 110)}`);
			return 0;
		}
		case "read": {
			const { name } = identify(room, flags.me, flags.cwd);
			const msgs = unread(room, name, { all: Boolean(flags.all) });
			if (!msgs.length) console.log(`(no new messages for ${name} in room ${room})`);
			else {
				for (const m of msgs) console.log(formatMessage(m));
				if (flags.ack) console.log(`(acked through #${ack(room, name, msgs.at(-1).seq)})`);
			}
			return 0;
		}
		case "wait": {
			const { name } = identify(room, flags.me, flags.cwd);
			const timeout = numberFlag(flags.timeout, 25, "--timeout"); // under every harness's shell-tool timeout (kiso: 30s)
			const nudge = statusNudge(room, name);
			if (nudge) {
				console.log(nudge);
				return 0;
			}
			const msgs = waitForMail(room, name, { timeoutMs: timeout * 1000, all: Boolean(flags.all) });
			if (!msgs.length) {
				console.log(`NO MAIL YET for ${name} after ${timeout}s. You are on the line only while this command runs.\nRun \`boss-call wait --timeout ${timeout}\` again NOW, with the same shell-tool timeout. Do not summarize, do not end your turn: an ended turn is a dropped line and nobody can call you back.`);
				return 0;
			}
			for (const m of msgs) console.log(formatMessage(m));
			console.log(`(acked through #${msgs.at(-1).seq}. Act on it and keep working; post a status when a piece is done and continue at once. Run \`boss-call wait\` only when nothing is left to do.)`);
			return 0;
		}
		case "ack": {
			const { name } = identify(room, flags.me, flags.cwd);
			console.log(`acked through #${ack(room, name)} for ${name}`);
			return 0;
		}
		case "post": {
			const { name } = identify(room, flags.me ?? flags.from, flags.cwd);
			const text = pos.length ? pos.join(" ") : readStdin();
			const rec = post(room, { from: name, to: flags.to, kind: flags.kind ?? "msg", text, ref: flags.ref });
			console.log(`posted #${rec.seq} ${rec.from} -> ${rec.to} [${rec.kind}]`);
			return 0;
		}
		case "serve":
			return serve({
				room, me: flags.me, cwd: flags.cwd, session: flags.session, bin: flags.kiso, profile: flags.profile, mode: flags.mode,
				envFile: flags["env-file"], poll: numberFlag(flags.poll, 30, "--poll", { min: 0.001 }), once: Boolean(flags.once),
			});
		case "peek-session":
			console.log(peek(pos[0] ?? "latest", { match: flags.match ?? null, lines: Number.parseInt(flags.lines ?? "12", 10), width: Number.parseInt(flags.width ?? "160", 10) }));
			return 0;
		default:
			die(`unknown command ${cmd} — try boss-call help`);
	}
}

function kisoOpts(flags) {
	const k = {};
	if (flags.profile) k.profile = flags.profile;
	if (flags["env-file"]) k.envFile = flags["env-file"];
	if (flags.kiso) k.bin = flags.kiso;
	if (flags.mode) k.mode = flags.mode;
	return Object.keys(k).length ? k : undefined;
}

function presence(hb) {
	if (!hb) return "never listened";
	const age = (Date.now() - Date.parse(hb.ts)) / 1000;
	if (hb.state === "waiting" && age < 180) return "listening";
	const ago = age < 90 ? `${Math.round(age)}s` : age < 5400 ? `${Math.round(age / 60)}m` : `${Math.round(age / 3600)}h`;
	return `${hb.state} ${ago} ago`;
}

function cmdWho({ room, me, cwd }) {
	const data = loadRoom(room);
	let who = { name: "?", role: "unknown" };
	try {
		who = identify(room, me, cwd);
	} catch {}
	console.log(`room    : ${room}  (${join(HOME, room)})`);
	console.log(`you     : ${who.name}  (${who.role})`);
	console.log(`boss    : ${data.boss?.name ?? "-"}${data.boss?.root ? `  root=${data.boss.root}` : ""}`);
	console.log("members :");
	for (const [n, m] of Object.entries(data.members)) console.log(`  ${n.padEnd(12)} ${m.root ?? ""}`);
	if (who.role === "member") console.log(`\nyou talk only to ${data.boss?.name}: \`boss-call post "..."\` goes there by default.`);
	else if (who.role === "boss") console.log("\nyou may `post --to <member>` or `--to all`; members can only reach you.");
	if (who.role !== "unknown") console.log("to put an agent session started here on the line, tell it: follow the boss-call skill");
	return 0;
}

function readStdin() {
	try {
		return require("node:fs").readFileSync(0, "utf8");
	} catch {
		return "";
	}
}

function readHelp() {
	const src = require("node:fs").readFileSync(fileURLToPath(import.meta.url), "utf8");
	return src.split("/**")[1].split("*/")[0].split("\n").map((l) => l.replace(/^\s*\*\s?/, "")).join("\n").trim();
}

import { createRequire } from "node:module";
const require = createRequire(import.meta.url);

try {
	process.exit(main(process.argv.slice(2)) ?? 0);
} catch (err) {
	if (err instanceof BossCallError) die(err.message);
	throw err;
}

File: boss-call/boss-call.py
#!/usr/bin/env python3
"""boss-call — one Boss, several members, one line each.

A room is a star: exactly one Boss and any number of members. A member talks
only to the Boss; the Boss talks to one member or to all. That is the whole
relationship, and the CLI enforces it rather than describing it.

Any agent that can run a shell command can be the Boss or a member: Claude
Code, Codex, kiso, a script. The harness reads one SKILL.md and calls this
command; `boss-call who` tells it which side it is on.

State lives outside every repository:

    ~/.model-boss/boss-call/<room>/
        room.json        {"boss": {"name", "root"?}, "members": {name: {"root"}}}
        messages.jsonl   {seq, ts, from, to, kind, text, ref?}   append-only
        cursors/<name>   last seq that member has acknowledged
        .lock            flock for append + cursor moves

A message is never an authorization. Spending money, pushing, merging and
deploying stay with the person at the terminal, whoever sent the message.
"""

from __future__ import annotations

import argparse
import fcntl
import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

HOME = Path(os.environ.get("BOSS_CALL_HOME", Path.home() / ".model-boss" / "boss-call"))
DEFAULT_ROOM = os.environ.get("BOSS_CALL_ROOM", "default")
KINDS = ("msg", "ask", "reply", "status")


# ----------------------------------------------------------------------------
# storage
# ----------------------------------------------------------------------------


def room_dir(room: str) -> Path:
    d = HOME / room
    d.mkdir(parents=True, exist_ok=True)
    (d / "cursors").mkdir(exist_ok=True)
    return d


class Locked:
    def __init__(self, room: str) -> None:
        self.path = room_dir(room) / ".lock"

    def __enter__(self) -> "Locked":
        self.fh = open(self.path, "a+")
        fcntl.flock(self.fh, fcntl.LOCK_EX)
        return self

    def __exit__(self, *exc: object) -> None:
        fcntl.flock(self.fh, fcntl.LOCK_UN)
        self.fh.close()


def load_room(room: str) -> dict:
    p = room_dir(room) / "room.json"
    if not p.exists():
        return {"boss": None, "members": {}}
    return json.loads(p.read_text())


def save_room(room: str, data: dict) -> None:
    (room_dir(room) / "room.json").write_text(json.dumps(data, indent=2, ensure_ascii=False) + "\n")


def read_messages(room: str) -> list[dict]:
    p = room_dir(room) / "messages.jsonl"
    if not p.exists():
        return []
    out = []
    for line in p.read_text().splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            out.append(json.loads(line))
        except json.JSONDecodeError:
            continue  # torn tail from a crash mid-write: skip, never guess
    return out


def get_cursor(room: str, name: str) -> int:
    p = room_dir(room) / "cursors" / name
    if not p.exists():
        return -1
    try:
        return int(p.read_text().strip())
    except ValueError:
        return -1


def set_cursor(room: str, name: str, seq: int) -> None:
    (room_dir(room) / "cursors" / name).write_text(f"{seq}\n")


# ----------------------------------------------------------------------------
# identity — who am I, and which side of the line
# ----------------------------------------------------------------------------


def _root_contains(root: str | None, here: Path) -> int:
    """Length of the matching root, or -1. Longest match wins."""
    if not root:
        return -1
    r = Path(root).expanduser().resolve()
    try:
        here.relative_to(r)
    except ValueError:
        return -1
    return len(str(r))


def identify(room: str, explicit: str | None, cwd: str | None = None) -> tuple[str, str]:
    """Returns (name, role) with role in {'boss', 'member'}.
    Order: --me, $BOSS_CALL_ME, then the registered root that contains the cwd."""
    data = load_room(room)
    boss = data.get("boss") or {}
    members = data.get("members", {})
    name = explicit or os.environ.get("BOSS_CALL_ME")
    if name:
        if boss.get("name") == name:
            return name, "boss"
        if name in members:
            return name, "member"
        sys.exit(f"boss-call: {name!r} is not in room {room!r} (boss: {boss.get('name')}, members: {', '.join(members) or 'none'})")
    here = Path(cwd or os.getcwd()).resolve()
    best: tuple[int, str, str] | None = None
    for n, m in members.items():
        k = _root_contains(m.get("root"), here)
        if k >= 0 and (best is None or k > best[0]):
            best = (k, n, "member")
    k = _root_contains(boss.get("root"), here)
    if k >= 0 and (best is None or k > best[0]):
        best = (k, boss["name"], "boss")
    if best is None:
        sys.exit(
            "boss-call: cannot tell who you are — pass --me <name>, set BOSS_CALL_ME, "
            "or run from inside a registered repo. Run `boss-call who --room " + room + "` to see the room."
        )
    return best[1], best[2]


# ----------------------------------------------------------------------------
# commands
# ----------------------------------------------------------------------------


def cmd_init(a: argparse.Namespace) -> int:
    data = load_room(a.room)
    if a.boss:
        name, _, root = a.boss.partition("=")
        data["boss"] = {"name": name, **({"root": str(Path(root).expanduser().resolve())} if root else {})}
    for spec in a.member:
        if "=" not in spec:
            sys.exit(f"--member wants name=/repo/root, got {spec!r}")
        name, root = spec.split("=", 1)
        data.setdefault("members", {})[name] = {"root": str(Path(root).expanduser().resolve())}
    if not data.get("boss"):
        sys.exit("boss-call: a room needs exactly one boss: --boss <name>[=/root]")
    save_room(a.room, data)
    return cmd_who(argparse.Namespace(room=a.room, me=data["boss"]["name"]))


def cmd_who(a: argparse.Namespace) -> int:
    data = load_room(a.room)
    boss = data.get("boss") or {}
    members = data.get("members", {})
    try:
        name, role = identify(a.room, a.me)
    except SystemExit:
        name, role = "?", "unknown"
    print(f"room    : {a.room}  ({room_dir(a.room)})")
    print(f"you     : {name}  ({role})")
    print(f"boss    : {boss.get('name', '-')}" + (f"  root={boss['root']}" if boss.get("root") else ""))
    print("members :")
    for n, m in members.items():
        print(f"  {n:12s} {m.get('root', '')}")
    if role == "member":
        print(f"\nyou talk only to {boss.get('name')}: `boss-call post \"...\"` goes there by default.")
    elif role == "boss":
        print("\nyou may `post --to <member>` or `--to all`; members can only reach you.")
    return 0


def cmd_post(a: argparse.Namespace) -> int:
    text = a.text if a.text is not None else sys.stdin.read()
    text = text.rstrip("\n")
    if not text.strip():
        sys.exit("boss-call: empty message")
    data = load_room(a.room)
    boss_name = (data.get("boss") or {}).get("name")
    sender, role = identify(a.room, a.sender)
    if role == "member":
        to = a.to or boss_name
        if to != boss_name:
            sys.exit(f"boss-call: a member talks only to the boss ({boss_name}); --to {to!r} refused")
    else:
        if not a.to:
            sys.exit("boss-call: the boss must say --to <member> or --to all")
        to = a.to
        if to != "all" and to not in data.get("members", {}):
            sys.exit(f"boss-call: {to!r} is not a member (members: {', '.join(data.get('members', {})) or 'none'})")
    with Locked(a.room):
        msgs = read_messages(a.room)
        seq = (msgs[-1]["seq"] + 1) if msgs else 0
        rec = {
            "seq": seq,
            "ts": datetime.now(timezone.utc).isoformat(timespec="seconds"),
            "from": sender,
            "to": to,
            "kind": a.kind,
            "text": text,
        }
        if a.ref is not None:
            rec["ref"] = a.ref
        with open(room_dir(a.room) / "messages.jsonl", "a") as fh:
            fh.write(json.dumps(rec, ensure_ascii=False) + "\n")
            fh.flush()
            os.fsync(fh.fileno())
    print(f"posted #{seq} {sender} -> {to} [{a.kind}]")
    return 0


def _addressed(m: dict, me: str) -> bool:
    return m["to"] in (me, "all") and m["from"] != me


def cmd_read(a: argparse.Namespace) -> int:
    me, role = identify(a.room, a.me)
    msgs = read_messages(a.room)
    cur = get_cursor(a.room, me)
    unread = [m for m in msgs if m["seq"] > cur and (a.all or _addressed(m, me))]
    if not unread:
        print(f"(no new messages for {me} in room {a.room!r})")
        return 0
    for m in unread:
        ref = f" ref={m['ref']}" if m.get("ref") else ""
        print(f"--- #{m['seq']} {m['ts']} {m['from']} -> {m['to']} [{m['kind']}]{ref}")
        print(m["text"])
    if a.ack:
        with Locked(a.room):
            set_cursor(a.room, me, msgs[-1]["seq"])
        print(f"(acked through #{msgs[-1]['seq']})")
    return 0


def cmd_ack(a: argparse.Namespace) -> int:
    me, _ = identify(a.room, a.me)
    msgs = read_messages(a.room)
    if msgs:
        with Locked(a.room):
            set_cursor(a.room, me, msgs[-1]["seq"])
        print(f"acked through #{msgs[-1]['seq']} for {me}")
    return 0


def cmd_status(a: argparse.Namespace) -> int:
    data = load_room(a.room)
    msgs = read_messages(a.room)
    boss = (data.get("boss") or {}).get("name", "-")
    names = [boss] + sorted(data.get("members", {}))
    print(f"room {a.room!r}: boss={boss}  {len(msgs)} messages")
    for n in names:
        cur = get_cursor(a.room, n)
        unread = [m for m in msgs if m["seq"] > cur and _addressed(m, n)]
        asks = [m for m in unread if m["kind"] == "ask"]
        last = max((m["ts"] for m in msgs if m["from"] == n), default="-")
        tag = "boss  " if n == boss else "member"
        print(f"  {tag} {n:12s} unread={len(unread):3d} asks={len(asks):2d} last-posted={last}")
    return 0


def cmd_tail(a: argparse.Namespace) -> int:
    for m in read_messages(a.room)[-a.n :]:
        first = m["text"].splitlines()[0] if m["text"] else ""
        print(f"#{m['seq']:4d} {m['ts']} {m['from']:>9s} -> {m['to']:<9s} [{m['kind']:6s}] {first[:110]}")
    return 0


# ----------------------------------------------------------------------------
# kiso session peek — the Boss reads the durable log instead of asking
# ----------------------------------------------------------------------------


def _sessions_dir() -> Path:
    return Path(os.environ.get("KISO_HOME", Path.home() / ".kiso")) / "sessions"


def _load_session(path: Path) -> list[dict]:
    events = []
    for line in path.read_text().splitlines():
        try:
            rec = json.loads(line)
        except json.JSONDecodeError:
            continue
        ev = rec.get("event")
        if isinstance(ev, dict):
            ev = dict(ev)
            ev["_runId"] = rec.get("runId")
            events.append(ev)
    return events


def _pick_session(spec: str, match: str | None) -> Path:
    d = _sessions_dir()
    files = sorted(d.glob("*.jsonl"), key=lambda p: p.stat().st_mtime, reverse=True)
    if spec != "latest":
        cands = [p for p in files if p.stem == spec or p.stem.endswith(spec)]
        if not cands:
            sys.exit(f"boss-call: no session matching {spec!r} in {d}")
        return cands[0]
    if match:
        for p in files:
            for ev in _load_session(p):
                if ev.get("type") == "user_input":
                    if match in str(ev.get("content", "")):
                        return p
                    break
        sys.exit(f"boss-call: no session whose first input contains {match!r}")
    if not files:
        sys.exit(f"boss-call: no sessions in {d}")
    return files[0]


def cmd_peek(a: argparse.Namespace) -> int:
    path = _pick_session(a.session, a.match)
    events = _load_session(path)
    if not events:
        print(f"{path.name}: empty")
        return 0
    last_run = events[-1]["_runId"]
    run = [e for e in events if e["_runId"] == last_run]
    inputs = [e for e in events if e.get("type") == "user_input"]
    terminal = next((e for e in reversed(run) if e.get("type") == "terminal"), None)
    last_in_idx = max((i for i, e in enumerate(events) if e.get("type") == "user_input"), default=-1)
    text = "".join(e.get("text", "") for e in events[last_in_idx + 1 :] if e.get("type") == "text_delta")
    decided = {e.get("decisionId") for e in events if e.get("type") in ("permission_decided", "permission_expired")}
    pending = [e for e in events if e.get("type") == "permission_requested" and e.get("decisionId") not in decided]
    receipts = {e.get("executionId") for e in run if e.get("type") in ("tool_execution_succeeded", "tool_execution_failed", "tool_execution_resolved")}
    uncertain = [e for e in run if e.get("type") == "tool_execution_started" and e.get("executionId") not in receipts]
    tool_calls = sum(1 for e in events if e.get("type") == "tool_call_end")
    mtime = datetime.fromtimestamp(path.stat().st_mtime).strftime("%Y-%m-%d %H:%M:%S")

    print(f"session {path.stem}  last-write {mtime}  events {len(events)}  turns {len(inputs)}  tool_calls {tool_calls}")
    if inputs:
        print(f"first input : {str(inputs[0].get('content', '')).replace(chr(10), ' ')[: a.width]}")
        print(f"last input  : {str(inputs[-1].get('content', '')).replace(chr(10), ' ')[: a.width]}")
    if terminal is None:
        state = "OPEN (no terminal in last run)"
        if pending:
            state += f" — WAITING ON A PERSON: {len(pending)} pending approval(s)"
        elif uncertain:
            state += f" — {len(uncertain)} uncertain execution(s) (started, no receipt)"
        else:
            state += " — running or interrupted"
    else:
        state = f"ended: {terminal.get('outcome', {}).get('kind', '?')}"
    print(f"state       : {state}")
    if text.strip():
        print("last text   :")
        for line in text.strip().splitlines()[-a.lines :]:
            print("  " + line[: a.width])
    return 0


# ----------------------------------------------------------------------------


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(prog="boss-call", description=__doc__.split("\n\n")[0])
    ap.add_argument("--room", default=DEFAULT_ROOM, help=f"room (default: $BOSS_CALL_ROOM or {DEFAULT_ROOM!r})")
    sub = ap.add_subparsers(dest="cmd", required=True)

    p = sub.add_parser("init", help="create/update a room: one boss, N members")
    p.add_argument("--boss", default=None, help="boss name, optionally name=/root")
    p.add_argument("--member", action="append", default=[], help="name=/repo/root (repeatable)")
    p.set_defaults(fn=cmd_init)

    p = sub.add_parser("who", help="who am I, who is the boss, who are the members")
    p.add_argument("--me", default=None)
    p.set_defaults(fn=cmd_who)

    p = sub.add_parser("post", help="send a message (a member's goes to the boss; the boss says --to)")
    p.add_argument("--me", "--from", dest="sender", default=None)
    p.add_argument("--to", default=None, help="boss only: a member name or 'all'")
    p.add_argument("--kind", choices=KINDS, default="msg")
    p.add_argument("--ref", default=None, help="what this is about: '#12', a commit, a file")
    p.add_argument("text", nargs="?", default=None, help="omit to read stdin")
    p.set_defaults(fn=cmd_post)

    p = sub.add_parser("read", help="print messages you have not acknowledged")
    p.add_argument("--me", default=None)
    p.add_argument("--all", action="store_true", help="include traffic not addressed to you")
    p.add_argument("--ack", action="store_true", help="mark what was printed as read")
    p.set_defaults(fn=cmd_read)

    p = sub.add_parser("ack", help="mark everything as read")
    p.add_argument("--me", default=None)
    p.set_defaults(fn=cmd_ack)

    p = sub.add_parser("status", help="unread and open asks per participant")
    p.set_defaults(fn=cmd_status)

    p = sub.add_parser("tail", help="last N messages, one line each")
    p.add_argument("-n", type=int, default=20)
    p.set_defaults(fn=cmd_tail)

    p = sub.add_parser("peek-session", help="summarise a kiso session from its durable log")
    p.add_argument("session", nargs="?", default="latest")
    p.add_argument("--match", default=None, help="with 'latest': newest session whose first input contains this")
    p.add_argument("--lines", type=int, default=12)
    p.add_argument("--width", type=int, default=160)
    p.set_defaults(fn=cmd_peek)

    a = ap.parse_args(argv)
    return a.fn(a)


if __name__ == "__main__":
    raise SystemExit(main())
